README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

[![Travis build
status](https://travis-ci.org/rich-iannone/ttt.svg?branch=master)](https://travis-ci.org/rich-iannone/ttt)
[![Coverage
status](https://codecov.io/gh/rich-iannone/ttt/branch/master/graph/badge.svg)](https://codecov.io/github/rich-iannone/ttt?branch=master)
[![CRAN
status](https://www.r-pkg.org/badges/version/ttt)](https://cran.r-project.org/package=ttt)

# Time Tracking Tool

Track the time taken to perform tasks with an input interface that is
fast and easy to use. We can add useful information alongside the task—a
label, a project identifier, info text—so that we have just enough
metadata. You can use a variety of shorthand input methods for quickly
defining dates and times (e.g., `yesterday at 2pm`, `2d ago at 18:30`,
etc.).

Tracked time events are stored on disk in a convenient location so that
we don’t have to explicitly manage the storage of tracked tasks.
Functions are available for inspecting the tracked tasks, returning
tibble objects. The stored data is easy to reason about and it’s human
readable: plain ol’ CSV files, one for each date where there are tracked
events.

**ttt**’s thinking: the troublesome task of tracking times takes *troppo
tempo*; the treatment? take, in total, the tiniest tidbit (a tincture,
truthfully) of time to tabularize those temporal *travails*.

## Adding Tracked Tasks

We can work with the following elements of a task:

  - `task (t)`: the task description
  - `start (s)`: the starting date and time (`YYYY-MM-DD HH:MM`)
  - `end (e)`: the ending date and time (`YYYY-MM-DD HH:MM`)
  - `label (l)`: an optional label to file the task under
  - `project (p)`: an optional project name/code for additional
    categorization
  - `info (i)`: an optional information note

Supppose we would like to track the following task:

| `element` | `content`                              |
| --------- | -------------------------------------- |
| `task:`   | `"Preparing presentation for Meetup"`  |
| `start:`  | `"2018-03-25 10:30"`                   |
| `end:`    | `"2018-03-25 14:20"`                   |
| `label`   | `"presentation"`                       |
| `project` | `"Meetup-2018-03-28"`                  |
| `info`    | `"Working on the introduction slides"` |

We can either use the `ttd()` function for drafting a tracked task (it
prints out a formatted tracked task entry in the console), or, the
`tt()` function which takes the tracked task information and commits it
to disk (still echoing the entry to the console).

Let’s make a time-tracking entry using the arguments of the `ttd()`
function:

``` r
ttd(
   task = "Preparing presentation for Meetup",
      s = "2018-03-25 10:30",
      e = "2018-03-25 14:20",
    lbl = "presentation",
   proj = "Meetup-2018-03-28",
   info = "Working on the introduction slides")
```

The following appears in the console to let us confirm this is correct.
Repeating the same call with `tt()` will commit the entry to disk.

    tracked task entry
    ------------------
    t:  Preparing presentation for Meetup
    s:  2018-03-25 10:30
    e:  2018-03-25 14:20
    l:  presentation
    p:  Meetup-2018-03-28
    i:  Working on the introduction slides

We can also use a simple shorthand for the same time-tracking entry.
Here we use a single string with single-letter element identifiers and
colons (e.g., `t:` for `task`, `p:` for `proj`). Each of these fields
are separated by a semicolon. Linebreaks are allowed and they are useful
for data entry. This yields the same console output:

``` r
ttd("
t: Preparing presentation for Meetup;
s: 2018-03-25 10:30; e: 2018-03-25 14:20;
l: presentation; p: Meetup-2018-03-28;
i: Working on the introduction slides
")
```

We can use all sorts of time shortcuts. For instance, if today is
`2018-03-25`, we can write the following for the `s:` and `e:` sections
and expect the same entry:

``` r
ttd("
t: Preparing presentation for Meetup;
s: 10:30a; e: 2:20p;
l: presentation; p: Meetup-2018-03-28;
i: Working on the introduction slides
")
```

Using times without dates assumes that the tracked time is for the
present day. When not using AM/PM markers (these can be `a`, `am`, `p`,
`pm`, and their capitalized forms), it is assumed that 24 hour time is
being used.

We can also use keywords such as `now` to insert the present date/time.
Events from days past can be logged using `yesterday at [time]` or `[n]d
ago at [time]`. Thus, a construction like `yesterday at 3:00p` is useful
for tracking tasks that began yesterday. Even something like `today
at 10:20am` will be interpreted the same as just using `10:20am`. Here’s
an example of this (written on `2018-03-26`):

``` r
ttd("
t: Preparing presentation for Meetup;
s: yesterday at 10:30a; e: yesterday at 2:20p;
l: presentation; p: Meetup-2018-03-28;
i: Working on the introduction slides
")
```

    tracked task entry
    ------------------
    t:  Preparing presentation for Meetup
    s:  2018-03-25 10:30
    e:  2018-03-25 14:20
    l:  presentation
    p:  Meetup-2018-03-28
    i:  Working on the introduction slides

When the draft entry looks satisfactory, re-run the same statement using
`tt()` instead of `ttd()`. You’ll receive the same print message in the
console along with a message stating that the tracked task was added.

``` r
tt("
t: Preparing presentation for Meetup;
s: yesterday at 10:30a; e: yesterday at 2:20p;
l: presentation; p: Meetup-2018-03-28;
i: Working on the introduction slides
")
```

    first tt of day added
    tracked task entry
    ------------------
    t:  Preparing presentation for Meetup
    s:  2018-03-25 10:30
    e:  2018-03-25 14:20
    l:  presentation
    p:  Meetup-2018-03-28
    i:  Working on the introduction slides

## Reading Tracked Tasks from Disk

Getting all the tracked tasks as a tibble can be done with the
`get_tt()` function:

``` r
get_tt()
```

    # A tibble: 1 x 8
      task                              s_d        s_t   e_d        e_t   lbl          proj   info   
      <chr>                             <chr>      <chr> <chr>      <chr> <chr>        <chr>  <chr>  
    1 Preparing presentation for Meetup 2018-03-31 10:30 2018-03-31 14:20 presentation Meetu… Workin…

We can narrow down the records returned by either using the `s` (start)
or `e` (end) arguments of the `get_tt()` function (both accept
ISO-formatted dates), or, by using **dplyr**’s `filter()` function
afterward.

Records are written to disk (one CSV file per day, naming format is
`[YYYY-MM-DD].csv`). Files are created and modified in the hidden `.ttt`
folder of the user’s Documents folder. The package will generate that
folder when necessary.

A few functions are provided to help with managing the files that *ttt*
creates:

  - `where_ttt_dir()`: provides the file path of the `.ttt` dir
  - `show_ttt_files()`: returns a vector of filenames produced though
    use of `tt()`
  - `delete_all_ttt_files()`: deletes all files in the `.ttt` directory

## Installation

You can install the development version of **ttt** from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("rich-iannone/ttt")
```

If you encounter a bug, have usage questions, or want to share ideas to
make this package better, feel free to file an
[issue](https://github.com/rich-iannone/ttt/issues).

## Code of Conduct

Please note that this project is released with a [Contributor Code of
Conduct](CODE_OF_CONDUCT.md). By participating in this project you agree
to abide by its terms.

## License

MIT © Richard Iannone


NEWS.md
# ttt 0.1.0

* New package ttt with main functions `tt()`, `ttd()`, and `get_tt()`


tests/testthat.R
library(testthat)
library(ttt)

test_check("ttt")


tests/testthat/test-ttd.R
context("Create a tracked time entry")

test_that("Creating a tracked time draft entry is possible", {

  # Capture output text emitted to console
  captured_output <-
    capture_output_lines(
      ttd(
        task = "Task for testthat",
        s = "2018-03-28 10:30",
        e = "2018-03-28 10:40",
        lbl = "testthat",
        proj = "Testing-ttt",
        info = "Draft tt entry"))

  # Expect certain lines to appear in the captured output
  expect_equal(
    captured_output[1],
    "tracked task entry")

  expect_equal(
    captured_output[2],
    "------------------")

  expect_equal(
    captured_output[3],
    "t:  Task for testthat")

  expect_equal(
    captured_output[4],
    "s:  2018-03-28 10:30")

  expect_equal(
    captured_output[5],
    "e:  2018-03-28 10:40")

  expect_equal(
    captured_output[6],
    "l:  testthat")

  expect_equal(
    captured_output[7],
    "p:  Testing-ttt")

  expect_equal(
    captured_output[8],
    "i:  Draft tt entry")
})


R/ttd.R
#' Draft a Tracked Task
#'
#' Use the syntax of the \code{tt()} function without
#' actually committing the task. This is useful for
#' practicing the shorthand of entering a tracked task
#' since the tracked-time event is echoed to the console.
#' @param ... a string that allows for a shorthand means to
#' track time for some task.
#' @param task the task description.
#' @param s the start time for the task. Must be
#' in the ISO date/time format of \code{YYYY-MM-DD HH:MM}.
#' @param e the end time for the task. Must be
#' in the ISO date/time format of \code{YYYY-MM-DD HH:MM}.
#' @param lbl a label to use for the event.
#' @param proj project information for the event.
#' @param info information pertaining to the event.
#' @examples
#' # Draft a tracked task using `ttd()`; it won't
#' # be committed to the ttt_dir but information
#' # on the tracked task will be echoed back--this
#' # is useful for ensuring that the information
#' # is correct before using `tt()`
#' ttd(
#'   task = "Description of the task",
#'   s = "2018-05-24 11:30",
#'   e = "2018-05-24 12:00",
#'   lbl = "label",
#'   proj = "project_name",
#'   info = "Any additional information")
#'
#' # We can also use the shorthand notation
#' ttd("
#' t: Description of the task;
#' s: 2018-05-24 11:30;
#' e: 2018-05-24 12:00;
#' l: label;
#' p: project_name;
#' i: Any additional information
#' ")
#' @export
ttd <- function(...,
                task = NULL,
                s = NULL,
                e = NULL,
                lbl = NULL,
                proj = NULL,
                info = NULL) {

  tt(... = ...,
     task = task,
     s = s,
     e = e,
     lbl = lbl,
     proj = proj,
     info = info,
     write = FALSE)
}


R/utils.R

ttt_dir_create <- function() {

  # Check for the existence of the `ttt_dir`
  if (dir.exists(path.expand("~/Documents/.ttt/")) == FALSE) {

    # Create the directory if it does not exist
    dir.create(path = path.expand("~/Documents/.ttt/"), recursive = TRUE)

    # Issue a message about the creation of the directory
    message(
      paste0(
        "The `.ttt` directory has been generated at ",
        where_ttt_dir()))
  }
}

ttt_dir_files <- function() {

  # Check for the existence of the `ttt_dir`
  if (dir.exists(path.expand("~/Documents/.ttt/"))) {

    # Create the directory if it does not exist
    list.files(path = path.expand("~/Documents/.ttt/"))
  } else {
    return("Path does not exist.")
  }
}

ttt_print_record <- function(tbl_row, title = NULL) {

  if (is.null(title)) {
    title <- "tracked task entry"
  }

  cat(
    paste0(
      title, "\n",
      "------------------\n",
      "t:  ", tbl_row[1]$task[1], "\n",
      "s:  ", tbl_row[2]$s_d[1], " ", tbl_row[3]$s_t, "\n",
      "e:  ", tbl_row[4]$e_d[1], " ", tbl_row[5]$e_t, "\n",
      "l:  ", tbl_row[6]$lbl[1], "\n",
      "p:  ", tbl_row[7]$proj[1], "\n",
      "i:  ", tbl_row[8]$info[1], "\n"))
}


R/ttt_fs.R

#' Get location of the ttt_dir
#'
#' Get the directory path for the hidden ttt directory.
#' @export
where_ttt_dir <- function() {

  path.expand("~/Documents/.ttt/")
}

#' Get a vector of files in ttt_dir
#'
#' Get a vector of all files in the ttt directory.
#' @param full_names an option for showing the
#' directory paths prepended to the file names.
#' full path
#' @export
show_ttt_files <- function(full_names = FALSE) {

  where_ttt_dir() %>% list.files(full.names = full_names)
}

#' Delete all files in ttt_dir
#'
#' Delete all of the files in the ttt directory.
#' @export
delete_all_ttt_files <- function() {

  file_count <-
    list.files(path = where_ttt_dir(), pattern = "*.csv") %>%
    length()

  invisible(
    file.remove(
      where_ttt_dir() %>%
      list.files(full.names = TRUE)))

  message(
    paste0(
      "All ", file_count, " CSV file(s) in `",
      where_ttt_dir(), "` have been removed"))
}


R/ttt_view.R
#' Get a table of tracked time within a range of dates
#'
#' Get a table of tracked time (as a \code{tbl_df}
#' object). The date range or number of days back
#' from the present day can be specified to filter
#' the table returned. Otherwise, all tracked time
#' records are provided.
#' @param s minimum day for which tasks are to be
#' collected. Date must be in ISO format:
#' \code{YYYY-MM-DD}.
#' @param e maximum day for which tasks are to be
#' collected. Date must be in ISO format:
#' \code{YYYY-MM-DD}.
#' @param last_n_days a number of days back from the
#' current day for which tasks are to be collected.
#' @examples
#' \dontrun{
#' # Show all tracked tasks as a tibble
#' get_tt()
#' #> A tibble: 3 x 8
#' #>   task                    s_d        s_t   e_d    e_t   lbl   proj   info
#' #>   <chr>                   <chr>      <chr> <chr>  <chr> <chr> <chr>  <chr>
#' #> 1 Description of the task 2018-05-24 11:30 2018-… 12:00 label proje… Any a…
#' #> 2 Description of the task 2018-05-24 12:00 2018-… 12:30 label proje… Any a…
#' #> 3 Description of the task 2018-05-24 12:30 2018-… 13:00 label proje… Any a…
#' }
#' @importFrom readr read_csv
#' @importFrom dplyr bind_rows mutate filter select
#' @export
get_tt <- function(s = NULL,
                   e = NULL,
                   last_n_days = NULL) {

  s_d <- s_d_date <- NULL

  ttt_file_list <- show_ttt_files(full_names = TRUE)

  for (i in seq(ttt_file_list)) {

    if (i == 1) {
      ttt_tbl <-
        ttt_file_list[i] %>%
        readr::read_csv(col_types = "cccccccc")
    }

    if (i > 1) {
      ttt_tbl <-
        dplyr::bind_rows(
          ttt_tbl,
          ttt_file_list[i] %>%
            readr::read_csv(col_types = "cccccccc"))
    }
  }

  if (is.null(last_n_days)) {

    if (!is.null(s)) {

      ttt_tbl <-
        ttt_tbl %>%
        dplyr::mutate(s_d_date = as.Date(s_d)) %>%
        dplyr::filter(s_d_date >= s) %>%
        dplyr::select(-s_d_date)
    }

    if (!is.null(e)) {

      ttt_tbl <-
        ttt_tbl %>%
        dplyr::mutate(s_d_date = as.Date(s_d)) %>%
        dplyr::filter(s_d_date <= e) %>%
        dplyr::select(-s_d_date)
    }

  } else {

    ttt_tbl <-
      ttt_tbl %>%
      dplyr::mutate(s_d_date = as.Date(s_d)) %>%
      dplyr::filter(s_d_date >= (Sys.Date() - as.difftime(last_n_days, units = "days"))) %>%
      dplyr::select(-s_d_date)
  }

  ttt_tbl
}


R/tt.R
#' Track Task (*tt*)
#'
#' Track a task by using \code{tt()}.
#' @param ... a string that allows for a shorthand means to
#' track time for some task.
#' @param task the task description.
#' @param s the start time for the task. Must be
#' in the ISO date/time format of \code{YYYY-MM-DD HH:MM}.
#' @param e the end time for the task. Must be
#' in the ISO date/time format of \code{YYYY-MM-DD HH:MM}.
#' @param lbl a label to use for the event.
#' @param proj project information for the event.
#' @param info information pertaining to the event.
#' @param write a logical value indicating whether the
#' tracked task entry should be written to disk.
#' @examples
#' \dontrun{
#' # Add a tracked task using `tt()`; it will
#' # be committed to the ttt_dir; information
#' # on the tracked task will be echoed back
#' tt(
#'   task = "Description of the task",
#'   s = "2018-05-24 12:30",
#'   e = "2018-05-24 13:00",
#'   lbl = "label",
#'   proj = "project_name",
#'   info = "Any additional information")
#' #> tracked task entry
#' #> ------------------
#' #> t:  Description of the task
#' #> s:  2018-05-24 11:30
#' #> e:  2018-05-24 12:00
#' #> l:  label
#' #> p:  project_name
#' #> i:  Any additional information
#'
#' # We can also use the shorthand notation
#' tt("
#' t: Description of the task;
#' s: 2018-05-24 11:30;
#' e: 2018-05-24 12:00;
#' l: label;
#' p: project_name;
#' i: Any additional information
#' ")
#' #> tracked task entry
#' #> ------------------
#' #> t:  Description of the task
#' #> s:  2018-05-24 11:30
#' #> e:  2018-05-24 12:00
#' #> l:  label
#' #> p:  project_name
#' #> i:  Any additional information
#' }
#' @importFrom stringr str_squish str_detect str_split str_remove
#' @importFrom stringr str_extract str_extract_all str_replace
#' @importFrom dplyr tibble bind_rows arrange
#' @importFrom readr read_csv write_csv
#' @importFrom purrr map
#' @importFrom rlang flatten_chr
#' @export
tt <- function(...,
               task = NULL,
               s = NULL,
               e = NULL,
               lbl = NULL,
               proj = NULL,
               info = NULL,
               write = TRUE) {

  x_in <- list(...)

  if (length(x_in) > 0) {
    x_in <- x_in[[1]]
  }

  # Validate input for `s`, if provided
  if (!is.null(s)) {

    s_present <- TRUE
    s <- s %>% stringr::str_squish()

    if (s %>% stringr::str_detect(
      pattern = "[1-9][0-9][0-9][0-9]-[0-1][0-9]-[0-3][0-9] [0-9]+\\:[0-5][0-9]")) {
      s_valid <- TRUE
    } else {
      s_valid <- FALSE
    }
  } else {
    s_present <- FALSE
  }

  # Validate input for `e`, if provided
  if (!is.null(e)) {

    e_present <- TRUE
    e <- e %>% stringr::str_squish()

    if (e %>% stringr::str_detect(
      pattern = "[1-9][0-9][0-9][0-9]-[0-1][0-9]-[0-3][0-9] [0-9]+\\:[0-5][0-9]")) {
      e_valid <- TRUE
    } else {
      e_valid <- FALSE
    }
  } else {
    e_present <- FALSE
  }

  if (inherits(x_in, "character")) {

    quick_input <-
      x_in %>%
      stringr::str_split(";") %>%
      purrr::map(.f = stringr::str_squish) %>%
      rlang::flatten_chr()

    # Parse any task given (t.*?:)
    if (any(quick_input %>% stringr::str_detect(pattern = "^t.*?\\:")) &
        is.null(task)) {

      # Get the vector component relating to a task (t:)
      component_t <-
        (quick_input %>%
           stringr::str_detect(pattern = "^t.*?\\:") %>%
           which())[1]

      # Extract the task text from the quick input
      task <-
        quick_input[component_t] %>%
        stringr::str_remove("^t.*?\\:") %>%
        stringr::str_squish()
    }

    # Parse any start time given (s.*?:)
    if (any(quick_input %>% stringr::str_detect(pattern = "^s.*?\\:")) &
        (!s_present || !s_valid)) {

      # Get the vector component relating to a start time (s:)
      component_s <-
        (quick_input %>%
           stringr::str_detect(
             pattern = "^s.*?\\:") %>%
           which())[1]

      # Extract the start time text from the quick input
      start <-
        quick_input[component_s] %>%
        stringr::str_remove("^s.*?\\:") %>%
        stringr::str_squish()

      # Get current date and time if `now` is used
      if (start %>%
          stringr::str_detect(
            pattern = ".*now.*")) {

        s_t <- format(Sys.time(), "%H:%M") %>% stringr::str_squish()
        s_d <- format(Sys.time(), "%F") %>% stringr::str_squish()
      }

      # Parse time if present
      if (start %>% stringr::str_detect(pattern = ".*\\:.*") |
          start %>% stringr::str_detect(pattern = "([1-9]|[1-9][0-9])[ ]*?[a-zA-Z][a-zA-Z]")) {

        if (start %>% stringr::str_detect(pattern = ".*\\:.*")) {

          s_t_n <-
            (quick_input[component_s] %>%
               stringr::str_remove("^s.*?\\:") %>%
               stringr::str_extract_all(pattern = "[0-9]+\\:[0-9][0-9]"))[[1]]

        } else if (start %>% stringr::str_detect(pattern = "([1-9]|[1-9][0-9])[ ]*?[a-zA-Z][a-zA-Z]")) {

          s_t_n <-
            paste0(
              quick_input[component_s] %>%
                stringr::str_remove("^s.*?\\:") %>%
                stringr::str_extract(pattern = "[0-9]*"),
              ":00")
        }

        if (start %>%
            stringr::str_detect(
              pattern = ".*pm$|.*PM$|.*p$|.*P$")) {

          s_hour_part <-
            (s_t_n %>%
               stringr::str_split(pattern = ":") %>%
               unlist())[1] %>%
            as.numeric()

          if (s_hour_part > 12) {
            s_t <- s_t_n

          } else {
            s_t <-
              paste0(
                (s_hour_part + 12), ":",
                (s_t_n %>%
                   stringr::str_split(pattern = ":") %>%
                   unlist())[2])
          }

        } else if (start %>%
                   stringr::str_detect(
                     pattern = ".*am$|.*AM$|.*a$|.*A$")) {

          s_t <- s_t_n

        } else {
          s_t <- s_t_n
        }
      }

      # Parse date if present
      if (start %>%
          stringr::str_detect(
            pattern = ".*2[0-9][0-9][0-9](-|//)[0-1][0-9](-|//)[0-3][0-9].*")) {

        s_year <-
          start %>%
          stringr::str_replace(
            pattern = ".*(2[0-9][0-9][0-9])(-|//)([0-1][0-9])(-|//)([0-3][0-9]).*",
            replacement = "\\1")

        s_month <-
          start %>%
          stringr::str_replace(
            pattern = ".*(2[0-9][0-9][0-9])(-|//)([0-1][0-9])(-|//)([0-3][0-9]).*",
            replacement = "\\3")

        s_day <-
          start %>%
          stringr::str_replace(
            pattern = ".*(2[0-9][0-9][0-9])(-|//)([0-1][0-9])(-|//)([0-3][0-9]).*",
            replacement = "\\5")

        s_d <- paste0(s_year, "-", s_month, "-", s_day)

      } else if (start %>% stringr::str_detect(pattern = ".*yesterday.*")) {

        s_d <-
          format(Sys.time() - as.difftime(1, units ="days"), "%F") %>%
          stringr::str_squish()

      } else if (start %>% stringr::str_detect(pattern = ".*[0-9]+?[ ]*?d ago.*")) {

        s_days_back <-
          start %>%
          stringr::str_extract("[0-9]+?[ ]*?d ago") %>%
          stringr::str_extract("[0-9]+?") %>%
          as.numeric()

        s_d <-
          format(Sys.time() - as.difftime(s_days_back, units ="days"), "%F") %>%
          stringr::str_squish()

      } else if (start %>% stringr::str_detect(pattern = ".*[0-9]+?[ ]*?d ahead.*")) {

        s_days_ahead <-
          start %>%
          stringr::str_extract("[0-9]+?[ ]*?d ahead") %>%
          stringr::str_extract("[0-9]+?") %>%
          as.numeric()

        s_d <-
          format(Sys.time() + as.difftime(s_days_ahead, units ="days"), "%F") %>%
          stringr::str_squish()

      } else if (start %>% stringr::str_detect(pattern = ".*\\-[0-9]+?[ ]*?d.*")) {

        s_days_back <-
          start %>%
          stringr::str_extract("\\-[0-9]+?[ ]*?d") %>%
          stringr::str_extract("[0-9]+?") %>%
          as.numeric()

        s_d <-
          format(Sys.time() - as.difftime(s_days_back, units ="days"), "%F") %>%
          stringr::str_squish()

      } else if (start %>% stringr::str_detect(pattern = ".*\\+[0-9]+?[ ]*?d.*")) {

        s_days_ahead <-
          start %>%
          stringr::str_extract("\\+[0-9]+?[ ]*?d") %>%
          stringr::str_extract("[0-9]+?") %>%
          as.numeric()

        s_d <-
          format(Sys.time() + as.difftime(s_days_ahead, units ="days"), "%F") %>%
          stringr::str_squish()

      } else {
        s_d <- format(Sys.time(), "%F") %>% stringr::str_squish()
      }
    } else {
      s_d <- NA_character_
      s_t <- NA_character_
    }

    # Parse any end time given (e.*?:)
    if (any(quick_input %>% stringr::str_detect(pattern = "^e.*?\\:")) &
        (!e_present || !e_valid)) {

      # Get the vector component relating to an end time (e:)
      component_e <-
        (quick_input %>%
           stringr::str_detect(
             pattern = "^e.*?\\:") %>%
           which())[1]

      # Extract the end time text from the quick input
      end <-
        quick_input[component_e] %>%
        stringr::str_remove("^e.*?\\:") %>%
        stringr::str_squish()

      # Get current date and time if `now` is used
      if (end %>%
          stringr::str_detect(
            pattern = ".*now.*")) {

        e_t <- format(Sys.time(), "%H:%M") %>% stringr::str_squish()
        e_d <- format(Sys.time(), "%F") %>% stringr::str_squish()
      }

      # Parse time if present
      if (end %>% stringr::str_detect(pattern = ".*\\:.*") |
          end %>% stringr::str_detect(pattern = "([1-9]|[1-9][0-9])[ ]*?[a-zA-Z][a-zA-Z]")) {

        if (end %>% stringr::str_detect(pattern = ".*\\:.*")) {

          e_t_n <-
            (quick_input[component_e] %>%
               stringr::str_remove("^e.*?\\:") %>%
               stringr::str_extract_all(pattern = "[0-9]+\\:[0-9][0-9]"))[[1]]

        } else if (end %>% str_detect(pattern = "([1-9]|[1-9][0-9])[ ]*?[a-zA-Z][a-zA-Z]")) {

          e_t_n <-
            paste0(
              quick_input[component_e] %>%
                stringr::str_remove("^e.*?\\:") %>%
                stringr::str_extract(pattern = "[0-9]*"),
              ":00")
        }

        if (end %>%
            stringr::str_detect(
              pattern = ".*pm$|.*PM$|.*p$|.*P$")) {

          e_hour_part <-
            (e_t_n %>%
               stringr::str_split(pattern = ":") %>%
               unlist())[1] %>%
            as.numeric()

          if (e_hour_part > 12) {
            e_t <- e_t_n

          } else {
            e_t <-
              paste0(
                (e_hour_part + 12), ":",
                (e_t_n %>%
                   stringr::str_split(pattern = ":") %>%
                   unlist())[2])
          }

        } else if (end %>%
                   stringr::str_detect(
                     pattern = ".*am$|.*AM$|.*a$|.*A$")) {

          e_t <- e_t_n

        } else {
          e_t <- e_t_n
        }
      }

      # Parse date if present
      if (end %>%
          stringr::str_detect(
            pattern = ".*2[0-9][0-9][0-9](-|//)[0-1][0-9](-|//)[0-3][0-9].*")) {

        e_year <-
          end %>%
          stringr::str_replace(
            pattern = ".*(2[0-9][0-9][0-9])(-|//)([0-1][0-9])(-|//)([0-3][0-9]).*",
            replacement = "\\1")

        e_month <-
          end %>%
          stringr::str_replace(
            pattern = ".*(2[0-9][0-9][0-9])(-|//)([0-1][0-9])(-|//)([0-3][0-9]).*",
            replacement = "\\3")

        e_day <-
          end %>%
          stringr::str_replace(
            pattern = ".*(2[0-9][0-9][0-9])(-|//)([0-1][0-9])(-|//)([0-3][0-9]).*",
            replacement = "\\5")

        e_d <- paste0(e_year, "-", e_month, "-", e_day)

      } else if (end %>% stringr::str_detect(pattern = ".*yesterday.*")) {

        e_d <-
          format(Sys.time() - as.difftime(1, units ="days"), "%F") %>%
          stringr::str_squish()

      } else if (end %>% stringr::str_detect(pattern = ".*[0-9]+?[ ]*?d ago.*")) {

        e_days_back <-
          end %>%
          stringr::str_extract("[0-9]+?[ ]*?d ago") %>%
          stringr::str_extract("[0-9]+?") %>%
          as.numeric()

        e_d <-
          format(Sys.time() - as.difftime(e_days_back, units ="days"), "%F") %>%
          stringr::str_squish()

      } else if (end %>% stringr::str_detect(pattern = ".*[0-9]+?[ ]*?d ahead.*")) {

        e_days_ahead <-
          end %>%
          stringr::str_extract("[0-9]+?[ ]*?d ahead") %>%
          stringr::str_extract("[0-9]+?") %>%
          as.numeric()

        e_d <-
          format(Sys.time() + as.difftime(e_days_ahead, units ="days"), "%F") %>%
          stringr::str_squish()

      } else if (end %>% stringr::str_detect(pattern = ".*\\-[0-9]+?[ ]*?d.*")) {

        e_days_back <-
          end %>%
          stringr::str_extract("\\-[0-9]+?[ ]*?d") %>%
          stringr::str_extract("[0-9]+?") %>%
          as.numeric()

        e_d <-
          format(Sys.time() - as.difftime(e_days_back, units ="days"), "%F") %>%
          stringr::str_squish()

      } else if (end %>% stringr::str_detect(pattern = ".*\\+[0-9]+?[ ]*?d.*")) {

        e_days_ahead <-
          end %>%
          stringr::str_extract("\\+[0-9]+?[ ]*?d") %>%
          stringr::str_extract("[0-9]+?") %>%
          as.numeric()

        e_d <-
          format(Sys.time() + as.difftime(e_days_ahead, units ="days"), "%F") %>%
          stringr::str_squish()

      } else {
        e_d <- format(Sys.time(), "%F") %>% stringr::str_squish()
      }
    } else {
      e_d <- NA_character_
      e_t <- NA_character_
    }

    # Parse any label given (l.*?:)
    if (any(quick_input %>% stringr::str_detect(pattern = "^l.*?\\:"))) {

      # Get the vector component relating to a label (l:)
      component_l <-
        (quick_input %>%
           stringr::str_detect(pattern = "^l.*?\\:") %>%
           which())[1]

      # Extract the label text from the quick input
      lbl <-
        quick_input[component_l] %>%
        stringr::str_remove("^l.*?\\:") %>%
        stringr::str_squish()
    }

    # Parse any project given (p.*?:)
    if (any(quick_input %>% stringr::str_detect(pattern = "^p.*?\\:"))) {

      # Get the vector component relating to a project (p:)
      component_p <-
        (quick_input %>%
           stringr::str_detect(pattern = "^p.*?\\:") %>%
           which())[1]

      # Extract the label text from the quick input
      proj <-
        quick_input[component_p] %>%
        stringr::str_remove("^p.*?\\:") %>%
        stringr::str_squish()
    }

    # Parse any information provided (i.*?:)
    if (any(quick_input %>% stringr::str_detect(pattern = "^i.*?\\:"))) {

      # Get the vector component relating to information (i:)
      component_i <-
        (quick_input %>%
           stringr::str_detect(pattern = "^i.*?\\:") %>%
           which())[1]

      # Extract the label text from the quick input
      info <-
        quick_input[component_i] %>%
        stringr::str_remove("^i.*?\\:") %>%
        stringr::str_squish()
    }
  }

  if (!is.null(s)) {

    if (s %>% stringr::str_detect(
      pattern = "^[0-9][0-9][0-9][0-9]-[0-1][0-9]-[0-3][0-9] [0-2][0-9]:[0-5][0-9]$")) {

      s_d <- (s %>% stringr::str_split(pattern = " ") %>% unlist())[1]

      s_t <- (s %>% stringr::str_split(pattern = " ") %>% unlist())[2]
    }
  }

  if (!is.null(e)) {

    if (e %>% stringr::str_detect(
      pattern = "^[0-9][0-9][0-9][0-9]-[0-1][0-9]-[0-3][0-9] [0-2][0-9]:[0-5][0-9]$")) {

      e_d <- (e %>% stringr::str_split(pattern = " ") %>% unlist())[1]

      e_t <- (e %>% stringr::str_split(pattern = " ") %>% unlist())[2]
    }
  }

  # Construct the table row
  tbl_row <-
    dplyr::tibble(
      task = ifelse(!is.null(task), task %>% as.character(), NA_character_),
      s_d = ifelse(exists("s_d"), s_d %>% as.character(), NA_character_),
      s_t = ifelse(exists("s_t"), s_t %>% as.character(), NA_character_),
      e_d = ifelse(exists("e_d"), e_d %>% as.character(), NA_character_),
      e_t = ifelse(exists("e_t"), e_t %>% as.character(), NA_character_),
      lbl = ifelse(!is.null(lbl), lbl %>% as.character(), NA_character_),
      proj = ifelse(!is.null(proj), proj %>% as.character(), NA_character_),
      info = ifelse(!is.null(info), info %>% as.character(), NA_character_))

  # Write to disk -----------------------------------------------------------

  if (write) {

    # Create the `ttt` dir in user's home if necessary
    ttt_dir_create()

    # The start date `s_d` is the reference date for CSV files generated;
    # if not provided use "2199-12-31"
    if (is.null(s) & !exists("s_d")) {
      file_name <- "2199-12-31.csv"
    } else if (!is.na(tbl_row$s_d)){
      file_name <- paste0(tbl_row$s_d, ".csv")
    }

    # If the file exists in the `ttt` dir, append row and sort by `s_d`;
    # otherwise, create a new file with a single row
    if (file_name %in% list.files(path = path.expand("~/Documents/.ttt/"))) {

      # Get the existing day info
      existing_day <-
        readr::read_csv(
          file = paste0(path.expand("~/Documents/.ttt/"), file_name),
          col_types = "cccccccc")

      previous_tt_n <- nrow(existing_day)

      # Bind new row to existing table
      updated_tt <-
        dplyr::bind_rows(
          existing_day, tbl_row) %>%
        dplyr::arrange(s_d, s_t)

      # Write revised table to `ttt` dir
      updated_tt %>%
        readr::write_csv(
          path = paste0(path.expand("~/Documents/.ttt/"), file_name))

      message(paste0("tt added, total tt: ", previous_tt_n + 1))

      ttt_print_record(tbl_row)

    } else {

      updated_tt <- tbl_row

      updated_tt %>%
        readr::write_csv(
          path = paste0(path.expand("~/Documents/.ttt/"), file_name))

      message("first tt of day added")

      ttt_print_record(tbl_row)
    }

  } else {

    ttt_print_record(tbl_row)
  }
}
